base/namedtuple.jl

```diff
@@ -0,0 +1,163 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+@generated function NamedTuple{names}(args...) where names
+    N = length(names)
+    if length(args) == N
+        Expr(:new, :(NamedTuple{names,$(Tuple{args...})}), Any[ :(args[$i]) for i in 1:N ]...)
+    else
+        :(throw(ArgumentError("wrong number of arguments to named tuple constructor")))
+    end
+end
+
+@generated function NamedTuple{names,T}(args...) where {names, T <: Tuple}
+    N = length(names)
+    types = T.parameters
+    if length(args) == N
+        Expr(:new, :(NamedTuple{names,T}), Any[ :(convert($(types[i]), args[$i])) for i in 1:N ]...)
+    else
+        :(throw(ArgumentError("wrong number of arguments to named tuple constructor")))
+    end
+end
+
+NamedTuple() = NamedTuple{()}()
+
+length(t::NamedTuple) = nfields(t)
+start(t::NamedTuple) = 1
+done(t::NamedTuple, iter) = iter > nfields(t)
+next(t::NamedTuple, iter) = (getfield(t, iter), iter + 1)
+endof(t::NamedTuple) = nfields(t)
+getindex(t::NamedTuple, i::Int) = getfield(t, i)
+getindex(t::NamedTuple, i::Symbol) = getfield(t, i)
+
+function getindex(t::NamedTuple, I::AbstractVector)
```

Comment:
Do we really need this method? It's never going to be type-stable and if I've followed @mbauman's recent indexing musings, this would be deprecated eventually anyway? (I could definitely be mis-understanding, but I think the idea is you'd do `nt.[I]` instead).

Something that I think would be more useful is something like:
```julia
getfield(t::NamedTuple, args...) = ...
```
where each arg would be an `Int` or `Symbol` and it would return a `NamedTuple` "subset" of the original NT. Is that currently do-able and be type stable? Would we need more constant folding/prop magic @vtjnash has been promising?

Reply:
Type stable logical indexing for tuples and named tuples also possible with typed bools and/or indexes.

Reply:
Coincidently, I just came across the `@pick` macro in the [IndexedTables.jl](https://github.com/JuliaComputing/IndexedTables.jl/blob/cc2d95afcf9e869b639fb2f0313eb4690381f957/src/utils.jl#L76) package that is pretty much what I was trying to describe here. It'd be great to have that kind of functionality supported/builtin.

Reply:
Agree that we can remove this `getindex` method.